src/execution/compiler/operator/cte_scan_leader_translator.cpp

```diff
@@ -0,0 +1,180 @@
+#include "execution/compiler/operator/cte_scan_leader_translator.h"
+#include "execution/compiler/codegen.h"
+#include "execution/compiler/compilation_context.h"
+#include "execution/compiler/function_builder.h"
+#include "execution/compiler/work_context.h"
+#include "parser/expression/constant_value_expression.h"
+//
+namespace terrier::execution::compiler {
+
+//
+CteScanLeaderTranslator::CteScanLeaderTranslator(const planner::CteScanPlanNode &plan,
+                                                 CompilationContext *compilation_context, Pipeline *pipeline)
+    : OperatorTranslator(plan, compilation_context, pipeline, brain::ExecutionOperatingUnitType::CTE_SCAN),
+      col_oids_var_(GetCodeGen()->MakeFreshIdentifier("col_oids")),
+      col_types_(GetCodeGen()->MakeFreshIdentifier("col_types")),
+      insert_pr_(GetCodeGen()->MakeFreshIdentifier("insert_pr")),
+      build_pipeline_(this, Pipeline::Parallelism::Serial) {
+  auto cte_type = GetCodeGen()->BuiltinType(ast::BuiltinType::Kind::CteScanIterator);
+  cte_scan_val_entry_ =
+      compilation_context->GetQueryState()->DeclareStateEntry(GetCodeGen(), plan.GetCTETableName() + "val", cte_type);
+  cte_scan_ptr_entry_ = compilation_context->GetQueryState()->DeclareStateEntry(
+      GetCodeGen(), plan.GetCTETableName() + "ptr", GetCodeGen()->PointerType(cte_type));
+
+  pipeline->LinkSourcePipeline(&build_pipeline_);
+  pipeline->UpdateParallelism(Pipeline::Parallelism::Serial);
+  compilation_context->Prepare(*(plan.GetChild(0)), &build_pipeline_);
+
+  std::vector<uint16_t> attr_sizes;
+  for (const auto &column : plan.GetTableSchema()->GetColumns()) {
+    attr_sizes.push_back(column.AttrSize());
+  }
+
+  auto offsets = storage::StorageUtil::ComputeBaseAttributeOffsets(attr_sizes, 0);
+
+  // Build the map from Schema columns to underlying columns
+  for (const auto &column : plan.GetTableSchema()->GetColumns()) {
```

Comment:
This code looks quite familiar and I feel it exists somewhere else, but maybe it doesn't.